### PR TITLE
OT Bridge - Initial implementation

### PR DESCRIPTION
**Pull request by johananl**
This is an initial implementation of the OpenTracing bridge following received feedback on https://github.com/open-telemetry/opentelemetry-python/pull/120. The bridge is an implementation of the OpenTracing `Tracer` class and includes several other OpenTracing subclasses required for implementing the tracer.

The following pieces are still not implemented:

- Inject/extract methods.
- Baggage items.

To run the unit tests, use the following:

```
git clone git@github.com:kinvolk/opentelemetry-python.git
cd opentelemetry-python
git checkout johananl/ot-shim
virtualenv /tmp/venv
source /tmp/venv/bin/activate
pip install tox
tox
```

To test actual tracing, use the following:

```
git clone git@github.com:kinvolk/opentelemetry-python.git
cd opentelemetry-python
git checkout johananl/ot-shim
virtualenv /tmp/venv
source /tmp/venv/bin/activate
cd opentracing-shim
python setup.py install
pip install opentelemetry-sdk
```

Then instrument an app:

```python
import time
from opentelemetry.sdk.trace import Tracer, export
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
from opentracingshim import create_tracer

otel_tracer = Tracer()
memory_exporter = InMemorySpanExporter()
span_processor = export.SimpleExportSpanProcessor(memory_exporter)
otel_tracer.add_span_processor(span_processor)

tracer = create_tracer(otel_tracer)

print('------------Started------------')

with tracer.start_active_span('parent') as parent:
    time.sleep(1)
    with tracer.start_active_span('child') as child:
        time.sleep(2)

for s in memory_exporter.get_finished_spans():
    print(s)
    print(s.get_context())

print('-------------Done--------------')
```

### CODE REVIEW

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    def tracer(self):
+        return self._tracer
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+        return self
+
+    def finish(self, finish_time=None):
+        self._otel_span.end()
+        # TODO: Handle finish_time. The OpenTelemetry API doesn't currently
```
Comment (krnowak) [inline opentracing-shim/src/opentracingshim/__init__.py]: It seems like it is possible now. But if you don't want to implement that right now, then maybe reword the TODO.

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    def close(self):
+        self._span.finish()
+        # TODO: Set active span on OpenTelemetry tracer.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class ScopeManagerWrapper(opentracing.ScopeManager):
+    def __init__(self, tracer: "TracerWrapper"):
+        # pylint: disable=super-init-not-called
```
Comment (krnowak) [inline opentracing-shim/src/opentracingshim/__init__.py]: I'm largely a python ignoramus, but I wonder if this thing shouldn't have a comment about why this linter warning should be ignored.

This is same for other cases of linter warning ignore clauses.

File: opentracing-shim/README.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
+OpenTelemetry Python API
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/README.rst]: There are some references to the API package in this README
- title
- pypi badge

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        # Return self for call chaining.
+        return self
+
+    def set_baggage_item(self, key, value):
+        return self
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        return None
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/src/opentracingshim/__init__.py]: Maybe `raise NotImplementedError`?
Apply as well for other missing places.

**Review by krnowak — COMMENTED** [review]
Two comments. But I'm not very well informed in the python stuff.

File: opentracing-shim/setup.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.4",
+        "Programming Language :: Python :: 3.5",
+        "Programming Language :: Python :: 3.6",
+        "Programming Language :: Python :: 3.7",
+    ],
+    description="OpenTracing shim for OpenTelemetry",
+    include_package_data=True,
+    long_description=open("README.rst").read(),
+    install_requires=["typing; python_version<'3.5'"],
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/setup.py]: `opentelemetry-api`?

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+    def finish(self, finish_time=None):
+        self._otel_span.end()
+        # TODO: Handle finish_time. The OpenTelemetry API doesn't currently
+        # support setting end time on a span and we cannot assume that all
+        # OpenTelemetry Tracer implementations have an `end_time` field.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/134
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/src/opentracingshim/__init__.py]: Should return `self` as other similar methods?

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    def start_span(
+        self,
+        operation_name=None,
+        child_of=None,
+        references=None,
+        tags=None,
+        start_time=None,
+        ignore_active_span=False,
+    ) -> SpanWrapper:
+        parent = child_of
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/src/opentracingshim/__init__.py]: Is this line really needed?

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            if not isinstance(references, list):
+                references = [references]
+            for ref in references:
+                span.add_link(ref.referenced_context.unwrap())
+
+        if tags:
+            for key, value in tags.items():
+                span.set_attribute(key, value)
+
+        span.start(start_time=start_time)
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/src/opentracingshim/__init__.py]: I'm not totally sure but I think both times are in different units:
Opentracing uses epoch as float in seconds: https://github.com/opentracing/opentracing-python/blob/7d2e62b5999d3cdec9a9c2720a34a1d11f425461/opentracing/span.py#L117
Opentelemetry uses epoch as int in nanoseconds https://github.com/open-telemetry/opentelemetry-python/blob/3b37675aa65b024eed15840f5b416d444213e5c2/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L252

**Review by mauriciovasquezbernal — COMMENTED** [review]
General implementation looks very good, just minor comments.

- What is the best approach for unimplemented methods? Do nothing or raise?
- I'm not totally sure I fully understand the details of https://github.com/open-telemetry/opentelemetry-python/issues/161. Will look into detail/ask you later on.

File: opentracing-shim/tests/test_shim.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+        self.assertIsNone(self.shim.active_span)
+
+        span.finish()
+
+    def test_explicit_span_finish(self):
+        """Test `finish()` method on `Span` objects."""
+
+        span = self.shim.start_span("TestSpan")
+
+        # Verify span hasn't ended.
```
Comment (johananl) [inline opentracing-shim/tests/test_shim.py]:
```suggestion
        # Verify the span hasn't ended.
```

File: opentracing-shim/src/opentracingshim/util.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+try:
+    time_ns = time.time_ns
+# Python versions < 3.7
+except AttributeError:
+
+    def time_ns():
+        return int(time.time() * 1e9)
+
+
+# A default event name to be used when logging events when a better event name
```
Comment (johananl) [inline opentracing-shim/src/opentracingshim/util.py]:
```suggestion
# A default event name to be used for logging events when a better event name
```

File: opentracing-shim/tests/test_shim.py [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        # Verify span has ended.
+        self.assertIsNotNone(scope.span.unwrap().end_time)
+
+        with self.shim.start_active_span(
+            "TestSpan", finish_on_close=False
+        ) as scope:
+            # Verify span hasn't ended.
+            self.assertIsNone(scope.span.unwrap().end_time)
+
+        # Verify span hasn't ended.
```
Comment (johananl) [inline opentracing-shim/tests/test_shim.py]:
```suggestion
        # Verify span hasn't ended after scope had been closed.
```

File: opentracing-shim/setup.py (hunk 5 above)
Comment (johananl) [reply opentracing-shim/setup.py]: Good point! Also `opentracing`.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 1 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: Good point! Implemented.

File: opentracing-shim/src/opentracingshim/__init__.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+    @property
+    def tracer(self):
+        return self._tracer
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+        return self
+
+    def finish(self, finish_time: float = None):
```
Comment (mauriciovasquezbernal) [inline opentracing-shim/src/opentracingshim/__init__.py]: Probably this should be `typing.optional[float]`, not sure why mypy doesn't comply.

related: https://github.com/open-telemetry/opentelemetry-python/pull/174#discussion_r331357011

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 6 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: Good catch! Done.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 12 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: You are right, however I'm considering removing type hints completely from the bridge since we cannot use any OpenTracing types. See https://gitter.im/open-telemetry/opentelemetry-python?at=5d9b74cd7aa5602ffc86c16c

Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: I've removed typing information from the package since IMO it's better to omit typing altogether than to include *partial* typing info. We can revisit this at a later stage.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 4 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: Good idea. Done.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 2 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: Excellent point. I've gone through similar cases and it turns out we *did* have to call `super().__init__()` on some cases. On other cases we don't even need to override the `__init__()` method from the base class. I'll push a cleanup in a bit.

In this specific case, though, all `super().__init__()` does is initialize a noop span and a noop scope so I don't see why we need to call it:

https://github.com/opentracing/opentracing-python/blob/7d2e62b5999d3cdec9a9c2720a34a1d11f425461/opentracing/scope_manager.py#L35-L36

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 7 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: It saves me from having to specify the `else` clause. I like this as it's more compact and less indented.
Please feel free to suggest an alternative (we need to allow `parent` to be `None`).

File: opentracing-shim/README.rst (hunk 3 above)
Comment (johananl) [reply opentracing-shim/README.rst]: Thanks. This was fixed already.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 8 above)
Comment (johananl) [reply opentracing-shim/src/opentracingshim/__init__.py]: Great catch. I've handled this by converting seconds to nanoseconds and vice versa.

File: opentracing-shim/src/opentracingshim/__init__.py (hunk 7 above)
Comment (mauriciovasquezbernal) [reply opentracing-shim/src/opentracingshim/__init__.py]: You're totally right. I didn't consider one of the `if`s.
